### ISSUE TITLE
gcc 5.2.0 compile error

### ISSUE THREAD

**Issue by helloleef** [facebook/folly#303]
I use gcc 5.2.0 compile folly . 

libtool: compile:  g++ -DHAVE_CONFIG_H -I./.. -pthread -I/usr/local/include -std=gnu++0x -g -O2 -MT detail/CacheLocality.lo -MD -MP -MF detail/.deps/CacheLocality.Tpo -c detail/CacheLocality.cpp  -fPIC -DPIC -o detail/.libs/CacheLocality.o
In file included from ./../folly/FBString.h:59:0,
                 from ./../folly/Conv.h:27,
                 from detail/CacheLocality.cpp:25:
./../folly/Traits.h:152:38: error: template argument 1 is invalid
   struct IsRelocatable<  __VA_ARGS__ > : std::true_type {};
                                      ^
./../folly/Traits.h:218:3: note: in expansion of macro 'FOLLY_ASSUME_RELOCATABLE'
   FOLLY_ASSUME_RELOCATABLE(__VA_ARGS__<T1, T2, T3>) }                   \
   ^
./../folly/Traits.h:424:1: note: in expansion of macro 'FOLLY_ASSUME_FBVECTOR_COMPATIBLE_3'
 FOLLY_ASSUME_FBVECTOR_COMPATIBLE_3(std::basic_string);
 ^
./../folly/Traits.h:168:48: error: template argument 1 is invalid
   struct has_nothrow_constructor<  __VA_ARGS__ > : ::boost::true_type {};
                                                ^
./../folly/Traits.h:221:5: note: in expansion of macro 'FOLLY_ASSUME_HAS_NOTHROW_CONSTRUCTOR'
     FOLLY_ASSUME_HAS_NOTHROW_CONSTRUCTOR(__VA_ARGS__<T1, T2, T3>) }
     ^
./../folly/Traits.h:424:1: note: in expansion of macro 'FOLLY_ASSUME_FBVECTOR_COMPATIBLE_3'
 FOLLY_ASSUME_FBVECTOR_COMPATIBLE_3(std::basic_string);
 ^
In file included from ./../folly/Conv.h:27:0,
                 from detail/CacheLocality.cpp:25:
./../folly/FBString.h:912:3: error: reference to 'basic_string' is ambiguous
   std::basic_string<Char> backend_;
   ^
In file included from ./../folly/FBString.h:59:0,
                 from ./../folly/Conv.h:27,
                 from detail/CacheLocality.cpp:25:
./../folly/Traits.h:244:9: note: candidates are: template<class T, class R, class A> class std::basic_string
   class basic_string;
         ^
In file included from /usr/local/include/c++/5.2.0/string:39:0,
                 from /usr/local/include/c++/5.2.0/stdexcept:39,
                 from /usr/local/include/c++/5.2.0/array:38,
                 from /usr/local/include/c++/5.2.0/tuple:39,
                 from /usr/local/include/c++/5.2.0/functional:55,
                 from ./../folly/detail/CacheLocality.h:23,
                 from detail/CacheLocality.cpp:17:
/usr/local/include/c++/5.2.0/bits/stringfwd.h:71:11: note:                 template<class _CharT, class _Traits, class _Alloc> class std::__cxx11::basic_string
     class basic_string;
           ^
In file included from ./../folly/Conv.h:27:0,
                 from detail/CacheLocality.cpp:25:
./../folly/FBString.h: In copy constructor 'folly::dummy_fbstring_core<Char>::dummy_fbstring_core(const folly::dummy_fbstring_core<Char>&)':
./../folly/FBString.h:871:9: error: class 'folly::dummy_fbstring_core<Char>' does not have any field named 'backend_'
       : backend_(another.backend_) {
         ^
./../folly/FBString.h: In constructor 'folly::dummy_fbstring_core<Char>::dummy_fbstring_core(const Char*, size_t)':
./../folly/FBString.h:874:9: error: class 'folly::dummy_fbstring_core<Char>' does not have any field named 'backend_'
       : backend_(s, n) {
         ^
./../folly/FBString.h: In member function 'void folly::dummy_fbstring_core<Char>::swap(folly::dummy_fbstring_core<Char>&)':
./../folly/FBString.h:877:5: error: 'backend_' was not declared in this scope
     backend_.swap(rhs.backend_);
     ^
./../folly/FBString.h: In member function 'const Char* folly::dummy_fbstring_core<Char>::data() const':
./../folly/FBString.h:880:12: error: 'backend_' was not declared in this scope
     return backend_.data();
            ^
./../folly/FBString.h: In member function 'Char* folly::dummy_fbstring_core<Char>::mutable_data()':
./../folly/FBString.h:884:14: error: 'backend_' was not declared in this scope
     return &*backend_.begin();
              ^
./../folly/FBString.h: In member function 'void folly::dummy_fbstring_core<Char>::shrink(size_t)':
./../folly/FBString.h:888:5: error: 'backend_' was not declared in this scope
     backend_.resize(size() - delta);
     ^
./../folly/FBString.h: In member function 'Char* folly::dummy_fbstring_core<Char>::expand_noinit(size_t)':
./../folly/FBString.h:892:5: error: 'backend_' was not declared in this scope
     backend_.resize(size() + delta);
     ^
./../folly/FBString.h: In member function 'void folly::dummy_fbstring_core<Char>::push_back(Char)':
./../folly/FBString.h:896:5: error: 'backend_' was not declared in this scope
     backend_.push_back(c);
     ^
./../folly/FBString.h: In member function 'size_t folly::dummy_fbstring_core<Char>::size() const':
./../folly/FBString.h:899:12: error: 'backend_' was not declared in this scope
     return backend_.size();
            ^
./../folly/FBString.h: In member function 'size_t folly::dummy_fbstring_core<Char>::capacity() const':
./../folly/FBString.h:902:12: error: 'backend_' was not declared in this scope
     return backend_.capacity();
            ^
./../folly/FBString.h: In member function 'void folly::dummy_fbstring_core<Char>::reserve(size_t)':
./../folly/FBString.h:908:5: error: 'backend_' was not declared in this scope
     backend_.reserve(minCapacity);
     ^

**Comment by zerotens** [comment]
Btw using Traits.h from https://github.com/facebook/folly/pull/214 fixes the problem

**Comment by 2underscores-vic** [comment]
The same issue when using GCC 6.1

**Comment by Orvid** [comment]
This should have been solved when #214 was merged, so I'll close this for now, could you open a separate issue for the GCC 6.1 error you're seeing?

**Comment by 2underscores-vic** [comment]
Actually it's the same error:

```
In file included from ./../folly/FBString.h:59:0,
                 from ./../folly/Conv.h:27,
                 from detail/CacheLocality.cpp:25:
./../folly/Traits.h:152:38: error: template argument 1 is invalid
   struct IsRelocatable<  __VA_ARGS__ > : std::true_type {};
                                      ^
./../folly/Traits.h:218:3: note: in expansion of macro 'FOLLY_ASSUME_RELOCATABLE'
   FOLLY_ASSUME_RELOCATABLE(__VA_ARGS__<T1, T2, T3>) }                   \
   ^~~~~~~~~~~~~~~~~~~~~~~~
./../folly/Traits.h:424:1: note: in expansion of macro 'FOLLY_ASSUME_FBVECTOR_COMPATIBLE_3'
 FOLLY_ASSUME_FBVECTOR_COMPATIBLE_3(std::basic_string);
 ^
./../folly/Traits.h:168:48: error: template argument 1 is invalid
   struct has_nothrow_constructor<  __VA_ARGS__ > : ::boost::true_type {};
                                                ^
./../folly/Traits.h:221:5: note: in expansion of macro 'FOLLY_ASSUME_HAS_NOTHROW_CONSTRUCTOR'
     FOLLY_ASSUME_HAS_NOTHROW_CONSTRUCTOR(__VA_ARGS__<T1, T2, T3>) }
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

Why close the issue when no release still available?

**Comment by Orvid** [comment]
I closed it because #214 was already merged which solved this issue for GCC 5.2, which is what this issue was specifically about.

Although yes, it is the same error, and the same root cause (the ABI of std::basic_string changing), yours is for GCC 6.1 and requires a different fix than the issue with GCC 5.2 does.
